### ISSUE TITLE
Moving Redox OS support to Tier 1

### ISSUE THREAD

**Issue by tertsdiepraam** [uutils/coreutils#5210]
For context, we recently made a [policy on platform support](https://uutils.github.io/coreutils/book/platforms.html). The gist of that is that a platform is Tier 1 if we have CI for it. I listed Redox there as a Tier 2 platform, because we don't have CI for it.

We want to support Redox OS, because it's an awesome project. uutils is also the default implementation of coreutils on Redox (at least it was when I last checked). However, we have probably accidentally broken support for Redox since https://github.com/uutils/coreutils/pull/2550. For example in https://github.com/uutils/coreutils/pull/5064, but there are probably many more failing tests to find. This is due to no fault of the authors of the PRs that broke the support, because it is simply impractical for us to keep compatibility without a Redox CI job. The longer we wait with setting this up, the harder it will be to fix.

I think the way to support this is as follows:

1. Investigate how bad the damage is, by building and testing the utils on Redox.
1. Fix some issues that break the entire build for all utils (e.g. the `terminal_size` issue mentioned below)
1. Figure out which subset of utils that already compiles and adapt this feature set accordingly:
https://github.com/uutils/coreutils/blob/1f081f3716b181cafbd0be9e2ef8d42a3c554330/Cargo.toml#L239-L245
1. Create a CI job to build, test and lint on Redox
1. Update our platform support page to make the support "official" :)
1. Open an issue to keep track of all the utils that should work but don't.
1. Then we can start accepting fixes to enable more utils.

I'm not sure what the best way is to add a CI job for Redox. We probably need to to set up a VM (similar to how we do FreeBSD).

Some things to explore:
- Are there examples of projects that already have Redox tests in their CI?
- What subset of utils make sense to support on Redox?

@jackpot51, hope you don't mind me pinging you here. I've tried to follow up on the outstanding issues from #2550:
- `filetime` seems to have made a release in the meantime here: https://github.com/alexcrichton/filetime/pull/72
- `termsize` has been removed from our dependency tree and replaced with `terminal_size` (https://github.com/uutils/coreutils/pull/3864). However, `terminal_size` makes no mention of Redox, so I suspect it won't work. We depend on this directly in `ls` and the `coreutils` multi-call binary, but also via `clap` (because we enabled `wrap_help` feature). So maybe we should first ask `terminal_size` to support redox. Otherwise we'll have to think about workarounds.
- Do you have any advice for a CI setup that works well in your opinion? Or any opinions/ideas/advice about this effort in general?

cc @sylvestre @cakebaker any opinions on this? Anything I missed?

**Comment by jackpot51** [comment]
I can make a PR for CI support. That is the best place to start since it will show what other items are broken.

**Comment by tertsdiepraam** [comment]
That would be great. Thanks!

**Comment by jackpot51** [comment]
Currently there are some issues with Redox's toolchain being based on 1.68. I'll need to update it again before trying this.

**Comment by ids1024** [comment]
https://github.com/uutils/coreutils/pull/5502 fixes the build on Redox. Not too many things needed to be changed.

A very easy way to do CI tests for different platforms is to just run a `cargo check` test, which for pure Rust code doesn't require a sysroot or linker and will catch most compile issues. Though for uutils this fails at the build script for `onig_sys`.

**Comment by tertsdiepraam** [comment]
Another reason to get rid of `onig` 😄 Good to hear it's not too difficult!

**Comment by ids1024** [comment]
`cargo check --target x86_64-unknown-redox --features feat_os_unix_redox` seems to work without a redox toolchain/sysroot if `expr` is commented out from `feat_common_core`, so `onig` seems to be the only issue with that currently.

**Comment by oech3** [comment]
Stopped by #9111

**Comment by ChrisDryden** [comment]
In other areas of the codebase the default regex utility was implemented with the regex with a TODO to say that it does not have the full GNU regex syntax support. Was wondering if we could add a feature flag that would be able to support using the regex crate in expr without the full GNU compatibility as a backstop for now to enable Redox to use it since to remove onig with a fully compatible implementation does not seem to be possible for a while or until a new dependency that has those capabilities is built

**Comment by jackpot51** [comment]
What is the regex crate missing?

**Comment by ChrisDryden** [comment]
I believe it's just back-references and look-around that's missing. I think all of the other available crates currently that support those two features are all just wrappers for c libraries such as onig

**Comment by oech3** [comment]
Is there any "prebuilt" redoxer we can use at CI?

**Comment by xtqqczze** [comment]
redox was removed from CI in https://github.com/uutils/coreutils/pull/10950 due to a tooling issue (https://github.com/uutils/coreutils/issues/10918)

**Comment by jackpot51** [comment]
We currently only have redoxer pre-built in the redoxos/redoxer docker image. Maybe cargo binstall could provide it?

**Comment by xtqqczze** [comment]
[`cargo binstall`](https://github.com/cargo-bins/cargo-binstall) uses binaries published to GitHub Releases or equivalent, these do not appear to exist: https://github.com/redox-os/redoxer/releases https://gitlab.redox-os.org/redox-os/redoxer/-/releases

**Comment by jackpot51** [comment]
I think, because the CI only does cargo check for Redox, that it is possible to just use upstream cargo.

**Comment by jackpot51** [comment]
If we need to provide binaries, github releases would not make sense as we only use github as a mirror.

**Comment by oech3** [comment]
Does it mean Ubuntu cargo?

**Comment by xtqqczze** [comment]
> If we need to provide binaries, github releases would not make sense as we only use github as a mirror.

I don't see binaries on GitLab either

**Comment by jackpot51** [comment]
You only need redoxer if you want to link binaries, as it sets up the cross compiler. If you just want to run `cargo check`, then you just have to have rustup and add the redox target.

**Comment by jackpot51** [comment]
No, we do not provide redoxer binaries outside of a docker image, for CI on Gitlab.

**Comment by jackpot51** [comment]
Other targets are compiling `cross` from source, why is compiling `redoxer` an issue?

**Comment by oech3** [comment]
>why is compiling redoxer an issue?

Because it is slow...

**Comment by xtqqczze** [comment]
Redox CI has been moved to check-only, due to https://github.com/uutils/coreutils/issues/10918#issue-3938879564

**Comment by jackpot51** [comment]
So, is it desired to fully link coreutils?

**Comment by oech3** [comment]
>Other targets are compiling cross from source

Really? Can we avoid it?

**Comment by jackpot51** [comment]
It is pretty clear reading the ci file that any target that has use-cross set to 'cross' would have to cargo install `cross`

**Comment by jackpot51** [comment]
I don't mind only doing a check on Redox, as it is highly likely to build provided there are no C dependencies

**Comment by jackpot51** [comment]
Once the next rustix release is done, I can submit a change that eliminates the need for redoxer.

**Comment by xtqqczze** [comment]
We don't know what broke compilation in CI, possibly libc related

**Comment by oech3** [comment]
>any target that has use-cross set to 'cross' would have to cargo install cross

How about zig build?

**Comment by jackpot51** [comment]
I'm confused, it seems that sccache is being added, so is that not being used to install `redoxer`?

**Comment by jackpot51** [comment]
https://github.com/uutils/coreutils/blob/ce08ea0fff61ad635e2e35249a395219e126bc0b/.github/workflows/CICD.yml#L657

**Comment by oech3** [comment]
caches are not reliable https://github.com/uutils/coreutils/issues/10347

**Comment by jackpot51** [comment]
For redox, I'd recommend to use `rustup target add x86_64-unknown-redox` followed by `cargo check --target x86_64-unknown-redox` and do not use redoxer. I don't think completely linking binaries matters much. I can help with this adjustment once the next rustix release happens if someone else does not take care of it.

**Comment by xtqqczze** [comment]
cc: @willnode

**Comment by xtqqczze** [comment]
@willnode Blocked on https://github.com/rust-lang/libc/pull/5530.

**Comment by xtqqczze** [comment]
It’s important that we have a check-only CI job for Redox:

```sh
cargo check --all-targets --features feat_os_unix_redox --target x86_64-unknown-redox
```

This would at least ensure that Redox-specific code is regularly checked by CI. For now, we can set `continue-on-error: true` so that existing errors don’t block CI.

**Comment by xtqqczze** [comment]
For now, this is blocked on:

- https://github.com/uutils/coreutils/pull/14620 (aa849c793d6bbee26b7afb097aacf80163472103)
- https://github.com/uutils/coreutils/issues/14704 (41abf5af50c67b2bb1f7b80326a7e74a7ce8ad19)
- https://github.com/uutils/coreutils/pull/14739 (aa849c793d6bbee26b7afb097aacf80163472103)
- libc 0.2.190 release (https://github.com/rust-lang/libc/pull/5530)
- https://github.com/uutils/coreutils/pull/14751
